File: Python/PlunyoCL/frontend/lexer.py
from enum import Enum
from typing import List, Dict

DIGITS: str = '0123456789'
LETTERS: str = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_'

class TokenType(Enum):
    # Literal Types
    Number = "Number"
    Identifier = "Identifier"
    
    # Keywords
    Set = "Set"
    Const = "Const"
    
    # Grouping & Operators
    BinaryOperator = "BinaryOperator"
    Equals = "Equals"
    Comma = ","
    Dot = "."
    Colon = ":"
    Semicolon = "Semicolon"
    
    L_Paren = "("
    R_Paren = ")"
    
    L_Brace = "{"
    R_Brace = "}"
    
    L_Bracket = '['
    R_Bracket = ']'

    EOF = "End Of File"
    
KEYWORDS: Dict[str, TokenType] = {
    "set": TokenType.Set,
    "const": TokenType.Const
}

class Token:
    def __init__(self, type: TokenType, value: str = ""):
        self.type: TokenType = type
        self.value: str = value

    def __repr__(self) -> str:
        return f"[ {self.type} : '{self.value}' ]"

def tokenize(source_code: str) -> List[Token]:
    tokens: List[Token] = []
    src: List[str] = list(source_code)

    while src:
        char: str = src.pop(0)

        if char.isspace(): 
            continue
        
        if char in '()+-*/%=^;}{:,][.':
            tokens.append(Token({
                '(': TokenType.L_Paren, 
                ')': TokenType.R_Paren,
                
                '{': TokenType.L_Brace,
                '}': TokenType.R_Brace,
                
                '[': TokenType.L_Bracket,
                ']': TokenType.R_Bracket,
                
                '=': TokenType.Equals,
                ';': TokenType.Semicolon,
                ':': TokenType.Colon,
                ',': TokenType.Comma,
                '.': TokenType.Dot,
                
                '+': TokenType.BinaryOperator,
                '-': TokenType.BinaryOperator, 
                '*': TokenType.BinaryOperator,
                '/': TokenType.BinaryOperator, 
                '%': TokenType.BinaryOperator,
                '^': TokenType.BinaryOperator
            }[char], char))
            
            
            # multichar tokens from here on out
        elif char in DIGITS:
            num: str = char
            while src and src[0] in DIGITS:
                num += src.pop(0)
            tokens.append(Token(TokenType.Number, float(num)))
        elif char in LETTERS:
            identifier: str = char
            while src and src[0] in LETTERS + DIGITS:
                identifier += src.pop(0)
                
            reserved: TokenType = KEYWORDS.get(identifier, None)
            tokens.append(Token(reserved if reserved is not None else TokenType.Identifier, identifier))
        else:
            raise ValueError(f"Unrecognized token: {char}")

    tokens.append(Token(TokenType.EOF))
    return tokens

File: Lua/PlunyoCL/frontend/lexer.lua
local Lexer = {}

-- token types
local TokenType = {
    NIL = "Nil", -- Literal types
    NUMBER = "Number",
    IDENTIFIER = "Identifier",

    VAR = "Var", -- Keywords

    BIN_OPERATOR = "Binary Operator", -- Grouping & Operators
    EQUALS = "Equals",
    L_PAREN = "Left Parenthesis",
    R_PAREN = "Right Parenthesis",

    QUOTE = "Quote",

    EOF = "E.O.F", -- End of file
}

TokenType.Lookup = {
    ["="] = TokenType.EQUALS,

    ["("] = TokenType.L_PAREN,
    [")"] = TokenType.R_PAREN,

    ["+"] = TokenType.BIN_OPERATOR,
    ["-"] = TokenType.BIN_OPERATOR,
    ["*"] = TokenType.BIN_OPERATOR,
    ["/"] = TokenType.BIN_OPERATOR,
    ["%"] = TokenType.BIN_OPERATOR,

    ["'"] = TokenType.QUOTE,
    ['"'] = TokenType.QUOTE,

    ["var"] = TokenType.VAR,
}

-- token "class"
local Token = {}
Token.__index = Token

function Token:new(type, value)
    return setmetatable({ type = type, value = value }, self)
end

Token.__tostring = function(self)
    return string.format('(Type: %s | Value: "%s")', self.type, self.value)
end

-- actual tokenizing
local keywords = {
    ["nil"] = TokenType.NIL,
    ["var"] = TokenType.VAR,
}

local function isAlpha(str)
    return str:match("^[A-Za-z]+$") ~= nil
end

local function isDigit(str)
    return str:match("^[0-9]+$") ~= nil
end

function Lexer.Tokenize(sourceCode)
    local tokens, src, index = {}, {}, 1

    local function at()
        return src[index]
    end

    local function eat()
        local current = src[index]
        index = index + 1
        return current
    end

    for i = 1, #sourceCode do
        src[i] = sourceCode:sub(i, i)
    end

    while index <= #src do
        local currentChar = at()

        if currentChar and currentChar:match("%s") then
            eat()
            goto continue
        end

        if currentChar and TokenType.Lookup[currentChar] then
            table.insert(
                tokens,
                Token:new(TokenType.Lookup[currentChar], currentChar)
            )

            eat()
        elseif currentChar and isDigit(currentChar) then
            local num = eat()

            while index <= #src and isDigit(at()) do
                num = num .. eat()
            end

            table.insert(tokens, Token:new(TokenType.NUMBER, num))
        elseif currentChar and isAlpha(currentChar) then
            local identifier = eat()

            while index <= #src and isAlpha(at()) do
                identifier = identifier .. eat()
            end

            table.insert(
                tokens,
                Token:new(
                    keywords[identifier] or TokenType.IDENTIFIER,
                    identifier
                )
            )
        elseif currentChar then
            error(string.format('Unexpected Token "%s"', currentChar))
        end

        ::continue::
    end

    return tokens
end

-- return the module so other files can require it
Lexer.TokenType = TokenType
Lexer.Token = Token

return Lexer
